crates/ide_assists/src/handlers/add_interpolation.rs

```diff
@@ -0,0 +1,410 @@
+                None => builder.replace(range, replace_text),
+            }
+        })
+    }
+}
+
+fn is_inside_escape_char(string_ast: &ast::String, offset: syntax::TextSize) -> bool {
+    let string_token = string_ast.syntax();
+    let string_range = string_token.text_range();
+    let mut inside_escape = false;
+    string_ast.escaped_char_ranges(&mut |range, _| {
+        let range = range + string_range.start();
+        if range.start() < offset && offset < range.end() {
+            inside_escape = true;
+        }
+    });
+    inside_escape
+}
+
+fn match_format_macro_arg_delims(string_ast: &ast::String) -> Option<Vec<SyntaxToken>> {
+    let string_token = string_ast.syntax();
+
+    let macro_call = ast::MacroCall::cast(string_token.ancestors().nth(1)?)?;
+    let macro_name = macro_call.path()?.segment()?.name_ref()?;
+
+    let format_arg_idx = match macro_name.text().as_str() {
+        "eprint" | "eprintln" | "format" | "panic" | "print" | "println" | "todo"
+        | "unimplemented" | "unreachable" => 0,
+        "assert" | "debug_assert" | "write" | "writeln" => 1,
+        "assert_eq" | "assert_ne" | "debug_assert_eq" | "debug_assert_ne" => 2,
+        _ => return None,
+    };
+
+    let token_tree = macro_call.token_tree()?;
+    let left_delim = token_tree.left_delimiter_token()?;
+    let right_delim = token_tree.right_delimiter_token()?;
+    let mut delimiters: Vec<SyntaxToken> = token_tree
+        .token_trees_and_tokens()
+        .flat_map(|not| not.as_token().cloned())
+        .filter(|token| token.kind() == T![,] || token == &left_delim || token == &right_delim)
+        .collect();
+
+    let format_arg_start = delimiters.get(format_arg_idx)?.text_range().end();
+    let format_arg_end = delimiters.get(format_arg_idx + 1)?.text_range().start();
+    let format_arg_range = TextRange::new(format_arg_start, format_arg_end);
+
+    if !format_arg_range.contains_range(string_token.text_range()) {
+        return None;
+    }
+
+    delimiters.drain(..format_arg_idx + 1);
+
+    if let Some(first_named) = first_named_argument(token_tree) {
+        delimiters.retain(|delim| delim.text_range().end() <= first_named.text_range().start());
+    }
+
+    Some(delimiters)
+}
+
+fn first_named_argument(token_tree: ast::TokenTree) -> Option<SyntaxToken> {
+    #[derive(Clone, Copy)]
+    enum NameState {
+        Identifier,
+        Equal,
+        Expression,
+    }
+
+    let mut name_state = NameState::Identifier;
+    let mut name_candiate = None;
+
+    for not in token_tree.token_trees_and_tokens() {
+        if let Some(token) = not.into_token() {
+            match (token.kind(), name_state) {
+                (syntax::SyntaxKind::IDENT, NameState::Identifier) => {
+                    name_candiate = Some(token);
+                    name_state = NameState::Equal;
+                }
+                (T![=], NameState::Equal) => {
+                    debug_assert!(name_candiate.is_some());
```

Comment:
This debug_assert seems fishy, what if we have something like `format!("{foo}", =)`, not really a valid call but still something that could be written. So this assert should go I think